// mdb-cli/helpers/index.js
'use strict';

var _require = require('./build-project'),
    buildProject = _require.buildProject;

var _require2 = require('./erase-project-directories'),
    eraseProjectDirectories = _require2.eraseProjectDirectories;

var _require3 = require('./create-package-json'),
    createPackageJson = _require3.createPackageJson;

var _require4 = require('./deserialize-object-from-file'),
    deserializeJsonFile = _require4.deserializeJsonFile;

var _require5 = require('./download-pro-starter'),
    downloadProStarter = _require5.downloadProStarter;

var _require6 = require('./fetch-products'),
    fetchProducts = _require6.fetchProducts;

var _require7 = require('./get-project-name'),
    getPackageName = _require7.getPackageName;

var _require8 = require('./get-sorted-products'),
    getSorted = _require8.getSorted;

var _require9 = require('./git-clone'),
    gitClone = _require9.gitClone;

var _require10 = require('./remove-folder'),
    removeFolder = _require10.removeFolder;

var _require11 = require('./save-token'),
    saveToken = _require11.saveToken;

var _require12 = require('./show-confirmation-prompt'),
    showConfirmationPrompt = _require12.showConfirmationPrompt;

module.exports = {
  buildProject: buildProject,
  eraseProjectDirectories: eraseProjectDirectories,
  createPackageJson: createPackageJson,
  deserializeJsonFile: deserializeJsonFile,
  downloadProStarter: downloadProStarter,
  fetchProducts: fetchProducts,
  getPackageName: getPackageName,
  getSorted: getSorted,
  gitClone: gitClone,
  removeFolder: removeFolder,
  saveToken: saveToken,
  showConfirmationPrompt: showConfirmationPrompt
};
